### PR TITLE
Add `Memory2` and `Memory3` package

### PR DESCRIPTION
**Pull request by jeongsoolee09**
## Description

Add `Memory2` and `Memory3` packages.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [X] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [X] Queries have been added for the following rules:
     - RULE-8-7-2
     - RULE-8-9-1
 - [ ] Queries have been modified for the following rules:
     - _rule number here_

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [ ] Yes
  - [X] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [ ] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: cpp/misra/test/rules/RULE-8-9-1/PointerComparedBetweenDifferentArrays.testref [hunk 1]
```diff
@@ -0,0 +1 @@
+cpp/common/test/rules/.ql
```
[the remark this answers is not in the document]
Comment (jeongsoolee09) [reply cpp/misra/test/rules/RULE-8-9-1/PointerComparedBetweenDifferentArrays.testref]: Resolved in [ad8864f](https://github.com/github/codeql-coding-standards/pull/1023/commits/ad8864f7421f7cc12c786d2f033f21fdfa8c4445).

File: rule_packages/cpp/Memory2.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        {
+          "description": "Pointer difference should be taken from pointers that belong to a same array.",
+          "kind": "path-problem",
+          "name": "Subtraction between pointers shall only be applied to ones that address elements of the same array",
+          "precision": "high",
+          "severity": "error",
+          "short_name": "PointerDifferenceTakenBetweenDifferentArrays",
+          "shared_implementation_short_name": "DoNotSubtractPointersAddressingDifferentArrays",
+          "tags": [
+            "scope/system"
```
Comment (MichaelRFairhurst) [inline rule_packages/cpp/Memory2.json]: Can you add the "correctness" tag to these?

**Review by MichaelRFairhurst — CHANGES_REQUESTED** [review]
Nice! Just one metadata comment only

File: rule_packages/cpp/Memory2.json (hunk 2 above)
Comment (jeongsoolee09) [reply rule_packages/cpp/Memory2.json]: Makes sense! I've also added it to `Memory3`, in [017f019](https://github.com/github/codeql-coding-standards/pull/1023/commits/017f019a0ccebcfb40c2d2379aac25da9dc38801).
